// @graphql-codegen/visitor-plugin-common/dist/esnext/base-visitor.js
import * as autoBind from 'auto-bind';
import { convertFactory } from './naming';
export class BaseVisitor {
    constructor(rawConfig, additionalConfig) {
        this._declarationBlockConfig = {};
        this._parsedConfig = {
            convert: convertFactory(rawConfig),
            typesPrefix: rawConfig.typesPrefix || '',
            namespacedImportName: rawConfig.namespacedImportName || null,
            externalFragments: rawConfig.externalFragments || [],
            addTypename: !rawConfig.skipTypename,
            nonOptionalTypename: !!rawConfig.nonOptionalTypename,
            ...(additionalConfig || {}),
        };
        this.scalars = {};
        Object.keys(this.config.scalars).forEach(key => {
            this.scalars[key] = this.config.scalars[key].type;
        });
        autoBind(this);
    }
    get config() {
        return this._parsedConfig;
    }
    convertName(node, options) {
        const useTypesPrefix = typeof (options && options.useTypesPrefix) === 'boolean' ? options.useTypesPrefix : true;
        return (useTypesPrefix ? this.config.typesPrefix : '') + this.config.convert(node, options);
    }
}
