// @graphql-codegen/typescript/dist/commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const graphql_1 = require("graphql");
const visitor_1 = require("./visitor");
exports.TsVisitor = visitor_1.TsVisitor;
const introspection_visitor_1 = require("./introspection-visitor");
exports.TsIntrospectionVisitor = introspection_visitor_1.TsIntrospectionVisitor;
tslib_1.__exportStar(require("./typescript-variables-to-object"), exports);
tslib_1.__exportStar(require("./visitor"), exports);
exports.plugin = (schema, documents, config) => {
    const visitor = new visitor_1.TsVisitor(schema, config);
    const printedSchema = graphql_1.printSchema(schema);
    const astNode = graphql_1.parse(printedSchema);
    const maybeValue = visitor.getMaybeValue();
    const visitorResult = graphql_1.visit(astNode, { leave: visitor });
    const introspectionDefinitions = includeIntrospectionDefinitions(schema, documents, config);
    const scalars = visitor.scalarsDefinition;
    return {
        prepend: [...visitor.getEnumsImports(), ...visitor.getScalarsImports(), maybeValue],
        content: [scalars, ...visitorResult.definitions, ...introspectionDefinitions].join('\n'),
    };
};
function includeIntrospectionDefinitions(schema, documents, config) {
    const typeInfo = new graphql_1.TypeInfo(schema);
    const usedTypes = [];
    const documentsVisitor = graphql_1.visitWithTypeInfo(typeInfo, {
        Field() {
            const type = graphql_1.getNamedType(typeInfo.getType());
            if (graphql_1.isIntrospectionType(type) && !usedTypes.includes(type)) {
                usedTypes.push(type);
            }
        },
    });
    documents.forEach(doc => graphql_1.visit(doc.content, documentsVisitor));
    const typesToInclude = [];
    usedTypes.forEach(type => {
        collectTypes(type);
    });
    const visitor = new introspection_visitor_1.TsIntrospectionVisitor(schema, config, typesToInclude);
    const result = graphql_1.visit(graphql_1.parse(graphql_1.printIntrospectionSchema(schema)), { leave: visitor });
    // recursively go through each `usedTypes` and their children and collect all used types
    // we don't care about Interfaces, Unions and others, but Objects and Enums
    function collectTypes(type) {
        if (typesToInclude.includes(type)) {
            return;
        }
        typesToInclude.push(type);
        if (graphql_1.isObjectType(type)) {
            const fields = type.getFields();
            Object.keys(fields).forEach(key => {
                const field = fields[key];
                const type = graphql_1.getNamedType(field.type);
                collectTypes(type);
            });
        }
    }
    return result.definitions;
}
exports.includeIntrospectionDefinitions = includeIntrospectionDefinitions;
