// @graphql-codegen/visitor-plugin-common/dist/commonjs/base-visitor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const autoBind = require("auto-bind");
const naming_1 = require("./naming");
class BaseVisitor {
    constructor(rawConfig, additionalConfig) {
        this._declarationBlockConfig = {};
        this._parsedConfig = {
            convert: naming_1.convertFactory(rawConfig),
            typesPrefix: rawConfig.typesPrefix || '',
            namespacedImportName: rawConfig.namespacedImportName || null,
            externalFragments: rawConfig.externalFragments || [],
            addTypename: !rawConfig.skipTypename,
            nonOptionalTypename: !!rawConfig.nonOptionalTypename,
            ...(additionalConfig || {}),
        };
        this.scalars = {};
        Object.keys(this.config.scalars).forEach(key => {
            this.scalars[key] = this.config.scalars[key].type;
        });
        autoBind(this);
    }
    get config() {
        return this._parsedConfig;
    }
    convertName(node, options) {
        const useTypesPrefix = typeof (options && options.useTypesPrefix) === 'boolean' ? options.useTypesPrefix : true;
        return (useTypesPrefix ? this.config.typesPrefix : '') + this.config.convert(node, options);
    }
}
exports.BaseVisitor = BaseVisitor;
