// @graphql-codegen/cli/dist/commonjs/init/plugins.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("./helpers");
const types_1 = require("./types");
exports.plugins = [
    {
        name: `TypeScript ${helpers_1.italic('(required by other typescript plugins)')}`,
        package: '@graphql-codegen/typescript',
        value: 'typescript',
        pathInRepo: 'typescript/typescript',
        available: hasTag(types_1.Tags.typescript),
        shouldBeSelected: tags => oneOf(tags, types_1.Tags.angular, types_1.Tags.stencil) || allOf(tags, types_1.Tags.typescript, types_1.Tags.react) || noneOf(tags, types_1.Tags.flow),
    },
    {
        name: `TypeScript Operations ${helpers_1.italic('(operations and fragments)')}`,
        package: '@graphql-codegen/typescript-operations',
        value: 'typescript-operations',
        pathInRepo: 'typescript/operations',
        available: tags => allOf(tags, types_1.Tags.browser, types_1.Tags.typescript),
        shouldBeSelected: tags => oneOf(tags, types_1.Tags.angular, types_1.Tags.stencil) || allOf(tags, types_1.Tags.typescript, types_1.Tags.react),
    },
    {
        name: `TypeScript Resolvers ${helpers_1.italic('(strongly typed resolve functions)')}`,
        package: '@graphql-codegen/typescript-resolvers',
        value: 'typescript-resolvers',
        pathInRepo: 'typescript/resolvers',
        available: tags => allOf(tags, types_1.Tags.node, types_1.Tags.typescript),
        shouldBeSelected: tags => noneOf(tags, types_1.Tags.flow),
    },
    {
        name: `Flow ${helpers_1.italic('(required by other flow plugins)')}`,
        package: '@graphql-codegen/flow',
        value: 'flow',
        pathInRepo: 'flow/flow',
        available: hasTag(types_1.Tags.flow),
        shouldBeSelected: tags => noneOf(tags, types_1.Tags.typescript),
    },
    {
        name: `Flow Operations ${helpers_1.italic('(operations and fragments)')}`,
        package: '@graphql-codegen/flow-operations',
        value: 'flow-operations',
        pathInRepo: 'flow/operations',
        available: tags => allOf(tags, types_1.Tags.browser, types_1.Tags.flow),
        shouldBeSelected: tags => noneOf(tags, types_1.Tags.typescript),
    },
    {
        name: `Flow Resolvers ${helpers_1.italic('(strongly typed resolve functions)')}`,
        package: '@graphql-codegen/flow-resolvers',
        value: 'flow-resolvers',
        pathInRepo: 'flow/resolvers',
        available: tags => allOf(tags, types_1.Tags.node, types_1.Tags.flow),
        shouldBeSelected: tags => noneOf(tags, types_1.Tags.typescript),
    },
    {
        name: `TypeScript Apollo Angular ${helpers_1.italic('(typed GQL services)')}`,
        package: '@graphql-codegen/typescript-apollo-angular',
        value: 'typescript-apollo-angular',
        pathInRepo: 'typescript/apollo-angular',
        available: hasTag(types_1.Tags.angular),
        shouldBeSelected: () => true,
    },
    {
        name: `TypeScript React Apollo ${helpers_1.italic('(typed components and HOCs)')}`,
        package: '@graphql-codegen/typescript-react-apollo',
        value: 'typescript-react-apollo',
        pathInRepo: 'typescript/react-apollo',
        available: tags => allOf(tags, types_1.Tags.react, types_1.Tags.typescript),
        shouldBeSelected: () => true,
    },
    {
        name: `TypeScript Stencil Apollo ${helpers_1.italic('(typed components)')}`,
        package: '@graphql-codegen/typescript-stencil-apollo',
        value: 'typescript-stencil-apollo',
        pathInRepo: 'typescript/stencil-apollo',
        available: hasTag(types_1.Tags.stencil),
        shouldBeSelected: () => true,
    },
    {
        name: `TypeScript MongoDB ${helpers_1.italic('(typed MongoDB objects)')}`,
        package: '@graphql-codegen/typescript-mongodb',
        value: 'typescript-mongodb',
        pathInRepo: 'typescript/mongodb',
        available: tags => allOf(tags, types_1.Tags.node, types_1.Tags.typescript),
        shouldBeSelected: () => false,
    },
    {
        name: `TypeScript GraphQL files modules ${helpers_1.italic('(declarations for .graphql files)')}`,
        package: '@graphql-codegen/typescript-graphql-files-modules',
        value: 'typescript-graphql-files-modules',
        pathInRepo: 'typescript/graphql-files-modules',
        available: tags => allOf(tags, types_1.Tags.browser, types_1.Tags.typescript),
        shouldBeSelected: () => false,
    },
    {
        name: `TypeScript GraphQL document nodes ${helpers_1.italic('(embedded GraphQL document)')}`,
        package: '@graphql-codegen/typescript-document-nodes',
        value: 'typescript-document-nodes',
        pathInRepo: 'typescript/document-nodes',
        available: tags => allOf(tags, types_1.Tags.typescript),
        shouldBeSelected: () => false,
    },
    {
        name: `Introspection Fragment Matcher ${helpers_1.italic('(for Apollo Client)')}`,
        package: '@graphql-codegen/fragment-matcher',
        value: 'fragment-matcher',
        pathInRepo: 'other/fragment-matcher',
        available: hasTag(types_1.Tags.browser),
        shouldBeSelected: () => false,
    },
];
function hasTag(tag) {
    return (tags) => tags.includes(tag);
}
function oneOf(list, ...items) {
    return list.some(i => items.includes(i));
}
function noneOf(list, ...items) {
    return !list.some(i => items.includes(i));
}
function allOf(list, ...items) {
    return items.every(i => list.includes(i));
}
