// @graphql-codegen/cli/dist/commonjs/init/targets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const fs_1 = require("fs");
const types_1 = require("./types");
async function guessTargets() {
    const pkg = JSON.parse(fs_1.readFileSync(path_1.resolve(process.cwd(), 'package.json'), {
        encoding: 'utf-8'
    }));
    const dependencies = Object.keys({
        ...pkg.dependencies,
        ...pkg.devDependencies
    });
    return {
        [types_1.Tags.angular]: isAngular(dependencies),
        [types_1.Tags.react]: isReact(dependencies),
        [types_1.Tags.stencil]: isStencil(dependencies),
        [types_1.Tags.browser]: false,
        [types_1.Tags.node]: false,
        [types_1.Tags.typescript]: isTypescript(dependencies),
        [types_1.Tags.flow]: isFlow(dependencies)
    };
}
exports.guessTargets = guessTargets;
function isAngular(dependencies) {
    return dependencies.includes('@angular/core');
}
function isReact(dependencies) {
    return dependencies.includes('react');
}
function isStencil(dependencies) {
    return dependencies.includes('@stencil/core');
}
function isTypescript(dependencies) {
    return dependencies.includes('typescript');
}
function isFlow(dependencies) {
    return dependencies.includes('flow');
}
