// @graphql-codegen/cli/dist/commonjs/utils/watcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const codegen_1 = require("../codegen");
const plugin_helpers_1 = require("@graphql-codegen/plugin-helpers");
const isValidPath = require("is-valid-path");
const isGlob = require("is-glob");
const debounce = require("debounce");
const logSymbols = require("log-symbols");
const debugging_1 = require("./debugging");
const logger_1 = require("./logger");
const path_1 = require("path");
const config_1 = require("../config");
const hooks_1 = require("../hooks");
function log(msg) {
    // double spaces to inline the message with Listr
    logger_1.getLogger().info(`  ${msg}`);
}
function emitWatching() {
    log(`${logSymbols.info} Watching for changes...`);
}
exports.createWatcher = (initialConfig, onNext) => {
    debugging_1.debugLog(`[Watcher] Starting watcher...`);
    let config = initialConfig;
    const files = [initialConfig.configFilePath].filter(a => a);
    const documents = plugin_helpers_1.normalizeInstanceOrArray(config.documents);
    const schemas = plugin_helpers_1.normalizeInstanceOrArray(config.schema);
    // Add schemas and documents from "generates"
    Object.keys(config.generates)
        .map(filename => plugin_helpers_1.normalizeOutputParam(config.generates[filename]))
        .forEach(conf => {
        schemas.push(...plugin_helpers_1.normalizeInstanceOrArray(conf.schema));
        documents.push(...plugin_helpers_1.normalizeInstanceOrArray(conf.documents));
    });
    if (documents) {
        documents.forEach(doc => {
            if (typeof doc === 'string') {
                files.push(doc);
            }
            else {
                files.push(...Object.keys(doc));
            }
        });
    }
    schemas.forEach((schema) => {
        if (isGlob(schema) || isValidPath(schema)) {
            files.push(schema);
        }
    });
    if (typeof config.watch !== 'boolean') {
        files.push(...plugin_helpers_1.normalizeInstanceOrArray(config.watch));
    }
    let watcher;
    const runWatcher = async () => {
        const chokidar = await Promise.resolve().then(() => require('chokidar'));
        let isShutdown = false;
        const debouncedExec = debounce(() => {
            if (!isShutdown) {
                codegen_1.executeCodegen(config)
                    .then(onNext, () => Promise.resolve())
                    .then(() => emitWatching());
            }
        }, 100);
        emitWatching();
        const ignored = [];
        Object.keys(config.generates)
            .map(filename => ({ filename, config: plugin_helpers_1.normalizeOutputParam(config.generates[filename]) }))
            .forEach(entry => {
            if (entry.config.preset) {
                const extension = entry.config.presetConfig && entry.config.presetConfig.extension;
                if (extension) {
                    ignored.push(path_1.join(entry.filename, '**', '*' + extension));
                }
            }
            else {
                ignored.push(entry.filename);
            }
        });
        watcher = chokidar.watch(files, {
            persistent: true,
            ignoreInitial: true,
            followSymlinks: true,
            cwd: process.cwd(),
            disableGlobbing: false,
            usePolling: true,
            interval: 100,
            binaryInterval: 300,
            depth: 99,
            awaitWriteFinish: true,
            ignorePermissionErrors: false,
            atomic: true,
            ignored,
        });
        debugging_1.debugLog(`[Watcher] Started`);
        const shutdown = async () => {
            isShutdown = true;
            debugging_1.debugLog(`[Watcher] Shutting down`);
            log(`Shutting down watch...`);
            watcher.close();
            hooks_1.lifecycleHooks(config.hooks).beforeDone();
        };
        // it doesn't matter what has changed, need to run whole process anyway
        watcher.on('all', async (eventName, path) => {
            hooks_1.lifecycleHooks(config.hooks).onWatchTriggered(eventName, path);
            debugging_1.debugLog(`[Watcher] triggered due to a file ${eventName} event: ${path}`);
            const fullPath = path_1.join(process.cwd(), path);
            if (eventName === 'change' && config.configFilePath && fullPath === config.configFilePath) {
                log(`${logSymbols.info} Config file has changed, reloading...`);
                const newParsedConfig = config_1.loadAndParseConfig(config.configFilePath);
                newParsedConfig.watch = config.watch;
                newParsedConfig.silent = config.silent;
                newParsedConfig.overwrite = config.overwrite;
                newParsedConfig.configFilePath = config.configFilePath;
                config = newParsedConfig;
            }
            debouncedExec();
        });
        process.once('SIGINT', shutdown);
        process.once('SIGTERM', shutdown);
    };
    // the promise never resolves to keep process running
    return new Promise((_, reject) => {
        codegen_1.executeCodegen(config)
            .then(onNext, () => Promise.resolve())
            .then(runWatcher)
            .catch(err => {
            watcher.close();
            reject(err);
        });
    });
};
